### ISSUE TITLE
[Deferred] SDL 0135 - "Push To Talk" key support

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#387]
Hello SDL community,

The review of "Push To Talk" key support" begins now and runs through February 6, 2018.  The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0135-PushToTalk-hardkey-support.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/387

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
A downside not mentioned is that this proposal would seem to allow developers to disable the head unit's built in infotainment voice recognition. The Push-to-talk hardkey should not conflict with the infotainment system's native system, or the option should be provided for the HU manufacturer to decide how to handle it (e.g. by sending the Push-to-talk key after the user says the app name).

**Comment by joeljfischer** [comment]
Would it work in the background? How would this work with multiple apps having registered for it?

**Comment by tsuzuki-uie** [comment]
@joeljfischer Thanks comment:
Certainly, this spec will be conflicted to "head unit's built in infotainment VR".
So, this spec mentioned that 
 if there is subscribed app then, HMI should not launch built in VR but send button event.
Condition is only "subscribed PUSH_TO_TALK or not".
After subscribed, built in VR will be disabled till unsubscribed.

"(e.g. by sending the Push-to-talk key after the user says the app name)"
Sure, such additinal behavior (condition and spec) should be able to be decided each OEMs (each HMI implementation/behavior). If such flexibility should be written in spec then, "OEM may implement ....", such description can be added.

> Would it work in the background? How would this work with multiple apps having registered for it?
It follows to OK/CUSTOM_BUTTON rule. So, FULL or LIMITED is required state. (In the case of HMI_STATE=BACKGROUND / NONE, the event will not be sent to app.)
OK/CUSTOM_BUTTON rule is implemented in core. However, If HMI (OEMs) would like to use built in VR (condition: COMMUNICATION apps are not FULL/LIMITED) then, HMI can hanlde PUSH_TO_TALK button in HMI, too (not send  OnButtonXX("PUSH_TO_TALK") to core, but handle it as trigger to launch built in VR). 
This behavior is not described to proposal. If you're agree, it would be added to proposal.

**Comment by joeljfischer** [comment]
I could see this being pretty confusing. The push to talk key would activate the head unit's system if no app is in the foreground and the app if an app *is* in the foreground, but there's no indication of that to the user. I'm not sure what a better UX would be for this situation, but I think it needs more discussion before the proposal is accepted as is.

**Comment by tsuzuki-uie** [comment]
@joeljfischer 
Sure. I hope OEMs and Voice Agent provider's opinion, before fixing spec.

**Comment by robinmk** [comment]
I too have concerns similar to what Joel has mentioned. Even if the voice agent app is in HMI state of FULL, the user maybe pressing PTT to access a native system function which may not be provided by the voice agent app (such as setting temperature or asking to call someone from the paired device). The logic on when to trigger the voice agent app vs system's native VR can lead to confusion.

**Comment by tsuzuki-uie** [comment]
Alternative, (ex. Alexa) (no new RPC)
0. App: addCommand(vrSynonym="Alexa")->callback:perfromAudioPassThru( TTS="may i help you?" )
1. UX: Push PTT -> launch HU native VR
2. UX: "Alexa"(this is recognized by HU native)
3. TTS: "may i help you?" (+AudioPassThru)
4. UX: "xxxxxx" (this is recorded as AudioPassThru, and recognized by App)

So, this proposal is only focus to remove step#2. 
If every one consider "demerit" which were already described by @joeljfischer is bigger than merit "can be skipped above step#2" then, this proposal should be deferred or rejected.
However, for voice agent service, simple and reliable step is important, I think.

**Comment by AndrewRMitchell** [comment]
Sorry for the late long feedback, I thought I clicked comment last night.

If this truly is a “Push to Talk” proposal (which I expect it is), then we probably need to expand it to explicitly cover both "Press to Talk" and "Hold to Talk” (which are different) use cases. This can be done simply by using the ButtonPressMode enum which already exists and is included in the API. ButtonPressMode could easily be mapped to LONG = Hold and SHORT = Press. Or we could extended the enum to also have a HOLD to allow short, long, and hold initiation for those HU's that support this. We will just need to include this on registration which this proposal does not currently do (and we should probably allow for registering all of them in addition to each individually or in any combination that makes sense for the app).

If the goal of this proposal is to only support the Press work flow (which I doubt), then in ordered to keep it consistent with the other API naming, it should be relabeled “PRESS_TO_TALK” to disambiguate its meaning.

Next, while declared out of scope for this proposal, it might be nice to formally define wake word support, even if not part of this proposal, as one that goes along with it. This proposal in one spot says this should already exist, and in another spot says it MAY exis. And while the steps for initiating are defined above (sort of proving it does exist), I do not believe it formally exists, and I would like to see the VA flow complete as soon as we can accomplish it. 

For Alexa we have both voice initiated and touch initiated. Touch initiated supports both tap and hold. So having all three formally defined would make the API complete from our standpoint. The voice initiated flow where the HU does the recognition, and then Alexa gets audio pass through isn't really one of our official workflows, so it will end up being treated as a form of PTT.

The only additional speech item again ignored from this proposal is the multi-turn flow where the user responds with additional information in response to a prompt. Again, I understand the initial gut reaction is to define it as out of scope, but even for the short press, is a short press in response to a prompt part of the continuation or is it an indication that the user wants a new interaction. Ignoring it doesn't make the problem any easier, but defining it makes for a far more consistent UX and happy users.

While declared out of scope for this proposal, there also seems to be discussion around the Mapping PTT Key to a specific App. I would see this a HU specific detail. My suggestion would be to keep the OEM specific use of the button until the user changes it, but allow the user to in a setting select from whatever Apps currently advertise their support of such a feature, with a fallback to HU control if and when there is nothing connected to the SLD interface. Of course the HU could revert back to NONE if the vendor wanted, but my guess is if the user doesn’t have their device connected, providing some functionality is better for the UX then providing none. This gives the OEM the option of what they want to do with the keys, and they could even add a different key from their internal voice flow if they wanted to.

Finally, if this is the full Push to Talk proposal, the OEM could map each Press and Hold to different Apps, or directly to the HU (even separating out short, long, and hold all as unique combinations). Again, there would need to be some UI to allow the user to set this up. But this would allow me to say assign a tap to the HU and a hold to an SDL App. Giving the user control to set things up in a way that make sense to them is a good thing. Overriding default behavior without user interaction is NOT the right thing to do (unless there are physical buttons specifically designed for SLD use) as users won't get that simply plugging a device in changes behavior of buttons.

In closing, I think this topic is quite a bit bigger then this one proposal (hopefully my comments have illustrated that, as I think they are longer then the proposal itself). To get it right really needs a workshop (IMHO), which I am sure that was not the goal of the author. I would put the goal to be nailing down the full VA workflow for SLD as opposed to just adding PTT, and I think that is something that would be of value for the whole SDLC community to get spelled out.

One final thought, I can see some OEM's would want to keep their VA on no matter what, while others might be willing to let an SDL VA handle everything (assuming they expose enough on the SDL). That is obviously up to the OEM and what model fits best with their needs.  I think this actually nicely fits with the work for Remote Control (or IAC) and to be done right, that would also be included, maybe in the same workshop.

**Comment by theresalech** [comment]
The Steering Committee voted to defer this proposal, as they agreed that more detailed discussion should occur in a workshop setting. This proposal will be deferred until a workshop takes place to determine next steps.

**Comment by theresalech** [comment]
In the workshop it was determined that the outcome of this proposal will be dependent on the outcome of the Pub/Sub Service Model proposal which will be entered in place of #376 and #385.

**Comment by theresalech** [comment]
This proposal is dependent on #599.  It will be updated accordingly on 2019-05-08 based on the outcome of #599.

**Comment by theresalech** [comment]
Proposal #599 was closed as inactive, so this proposal will be closed as well.  The issue will remain in a `deferred` status and unlocked so the author can notify the Project Maintainer when it is ready to be revisited.
